_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+
+In this post, I'll walk through how the OAuth 2.0 Device Authorization Grant works and build a basic Java console application!
+
+If you want to skip the post and jump directly into the code, you can find it on oktadev/okta-device-grant-java-example[GitHub].
+
+**Prerequisites:**
+
+- An Okta organization - (Sign up for free using your https://developer.okta.com/signup/[browser], or the https://cli.okta.com/[Okta CLI])
+- https://adoptium.net/[Java 11+] (Tested with 11, 17, & 18)
+- https://www.jbang.dev/download/[JBang] (v0.92.2)
+
+toc::[]
+
+== How does the OAuth 2.0 Device Authorization Grant work?
+
+At a high level, the flow works like this:
+
+1. A device or application initiates the flow `POST /v1/device/authorize`
+2. The IdP responds with a "device code", "user code", and a URL.
+3. The user is instructed to visit a URL,  enter the "user code", and authenticate.
+4. The application polls the IdP using the "device code" until the user has logged in.
+5. The IdP response with an access token.
+
+image::{% asset_path 'blog/java-cli-device-grant/flow-diagram.jpg' %}[alt=Device grant flow diagram,width=800,align=center]
+
+////
+Tweak the diagram on https://mermaid.live/ with the following content
+flowchart TD
+    A[Start Login] -- POST to /v1/device/authorize --> B{Can open Browser?};
+    B -- Yes --> C[Open Browser to \n response.verification_uri_complete];
+    B -- No --> D{show QR Code?};
+    D -- Yes --> E[Create QR Code with \n response.verification_uri_complete];
+    D -- No --> F[Instruct user visit \n response.verification_uri and \n input response.user_code];
+    C --> G[Wait for user to \n login on other device];
+    E --> G;
+    F --> G;
+    G[sleep for response.interval] --> H[POST to /v1/token with \n response.device_code];
+    H --> I{Is Success};
+    I -- Yes --> J(User Is authenticated);
+    I -- No --> G;
+////
+
+NOTE: This example will work the same with any Identity Provider (IdP) that supports the flow, but you may need to tweak the URLs.
+
+== OAuth 2.0 from the command line
+
+You might not think of your terminal as an "input restricted device", but that still doesn't mean you should start typing a website's passwords; there are a few problems related to this:
+
+* You should ONLY enter a password into the website it was created for.
+* Your terminal doesn't support the same factors your browser does.
+* You may be ssh'd into another computer.
+
+NOTE: The Authorization Device Grant requires a user. If you are trying to call APIs from an automated process or other server-to-server communication, check out the link:/blog/2021/05/05/client-credentials-spring-security[OAuth 2.0 Client Credential Flow].
+
+== Setup Okta
+
+First, you need to create an Okta application that supports the device grant.
```

Comment:
```suggestion
First, you need to create an Okta application that supports the Device Grant.
```
a bit unsure about the capitalization, but aiming. to be consistent. Wish me luck!